### PR TITLE
util: Add UTIL_LISTIFY test cases

### PR DESCRIPTION
**Pull request by finikorg**
Added tests for UTIL_LISTIFY and corrected test name.

### CODE REVIEW

**Review by SebastianBoe — CHANGES_REQUESTED** [review]
An AFAICT equally[0] powerful mechanism is already present.

To reduce the amount of infrastructure, I believe it would be better to re-use what we have.

https://github.com/zephyrproject-rtos/zephyr/commit/4dc9f868bc437fd9befede8b7435b81c15cc74f2

[0] Actually, more flexible, as it doesn't require ';' to separate each element and is therefore more re-usable than LOOP.

File: include/misc/util.h [hunk 1]
```diff
@@ -576,4 +576,22 @@ static inline s64_t arithmetic_shift_right(s64_t value, u8_t shift)
...
+ * following construction might be used:
+ *
+ * UTIL_EVAL(UTIL_DEFER(LOOP) (statement, number))
+ */
+
+#define loop_0(_fn, n)
+#define loop_1(_fn, n) _fn(0)
+#define loop_2(_fn, n) loop_1(_fn, n); _fn(1)
+#define loop_3(_fn, n) loop_2(_fn, n); _fn(2)
+#define loop_4(_fn, n) loop_3(_fn, n); _fn(3)
```
Comment (andyross) [inline include/misc/util.h]: Obviously there's going to be an eternal argument about how many defined iteration counts are useful or practical to implement, etc...

Still, I feel confident that four is not going to be enough. :)

File: include/misc/util.h [hunk 2]
```diff
@@ -576,4 +576,22 @@ static inline s64_t arithmetic_shift_right(s64_t value, u8_t shift)
...
 	 (fixed_arg, x, ##__VA_ARGS__)}
 
+/*
+ * LOOP is used for a number of statements with incremented argument, can
+ * be used for defining number of serial ports or other devices
+ *
+ * Note: When taking an argument configuration parameters (CONFIG_NUM_OF_xxx)
+ * following construction might be used:
+ *
+ * UTIL_EVAL(UTIL_DEFER(LOOP) (statement, number))
```
Comment (andyross) [inline include/misc/util.h]: Seems like this "note" needs a little more explanation.  Recursive preprocessor expansion is deep voodoo.  I'd suggest a full, concrete example here and clearer text like for example:

    The n argument MUST be a literal decimal number in the range [1..4],
    so if you want it to be expanded from another preprocessor macro you
    must pre-expand the expression like this:

    UTIL_EVAL(UTIL_DEVER(LOOP)(MY_MACRO_TO_EXPAND, CONFIG_MY_NUMBER_OF_ITERATIONS)

File: include/misc/util.h [hunk 3]
```diff
@@ -576,4 +576,22 @@ static inline s64_t arithmetic_shift_right(s64_t value, u8_t shift)
...
+ */
+
+#define loop_0(_fn, n)
+#define loop_1(_fn, n) _fn(0)
+#define loop_2(_fn, n) loop_1(_fn, n); _fn(1)
+#define loop_3(_fn, n) loop_2(_fn, n); _fn(2)
+#define loop_4(_fn, n) loop_3(_fn, n); _fn(3)
+
+#define LOOP(fun, n) loop_##n(fun, n)
```
Comment (andyross) [inline include/misc/util.h]: Seems like this should be UTIL_LOOP to conform with existing namespace.  Also maybe STATIC_LOOP or something might better convey the idea here that it's a compile-time expansion?

**Review by andyross — APPROVED** [review]
Seems really clever.  Random nitpikery.

**Comment by nordic-krch** [comment]
there is `UTIL_LISTIFY` macro. It looks that it's doing the same.

**Comment by finikorg** [comment]
> An AFAICT equally[0] powerful mechanism is already present.
> 
> To reduce the amount of infrastructure, I believe it would be better to re-use what we have.
> 
> [4dc9f86](https://github.com/zephyrproject-rtos/zephyr/commit/4dc9f868bc437fd9befede8b7435b81c15cc74f2)
> 
> [0] Actually, more flexible, as it doesn't require ';' to separate each element and is therefore more re-usable than LOOP.

Thanks for notice, I will try to use `UTIL_LISTIFY`, could not understand what it does by description.

**Comment by finikorg** [comment]
> > could not understand what it does by description.
> 
> Please do post a PR if you can find a better description.

I started to use `UTIL_LISTIFY` but it requires to add  `;` and I have warnings

**Comment by SebastianBoe** [comment]
`but it requires to add `;``

Yes it does, but this is an acceptable price to pay to not have to maintain two macros that nearly do the same thing.

```
LOOP()
LOOP_WITHOUT_SEMICOLON()

```

What kind of warnings?

**Comment by finikorg** [comment]
> What kind of warnings?
Check warnings in this PR.

**Comment by SebastianBoe** [comment]
> Check warnings in this PR.

Not sure how, "Details" doesn't give any more details.

Anyhow, these macro's often violate the static analyzer's rules, I believe a maintainer usually ignores the rule and merges it anyway.

**Review by SebastianBoe — APPROVED** [review]

**Comment by SebastianBoe** [comment]
This error seems reasonable to me:

`-:13: ERROR:MULTISTATEMENT_MACRO_USE_DO_WHILE: Macros with multiple statements should be enclosed in a do - while loop`

do { i += x; } while(0);

**Comment by finikorg** [comment]
> This error seems reasonable to me:
> 
> `-:13: ERROR:MULTISTATEMENT_MACRO_USE_DO_WHILE: Macros with multiple statements should be enclosed in a do - while loop`
> 
> do { i += x; } while(0);

-:10: WARNING:SINGLE_STATEMENT_DO_WHILE_MACRO: Single statement macros should not use a do {} while (0) loop
#10: FILE: tests/misc/util/src/main.c:63:
+#define INC(x, _) do { i += x; } while (0);

**Review by nordic-krch — APPROVED** [review]

**Comment by SebastianBoe** [comment]
Can't catch a break ...

I don't know how to resolve the warnings.

**Comment by nashif** [comment]
@SebastianBoe 
> Not sure how, "Details" doesn't give any more details.

find first comment by zephyrbot, browse through the history of the comment.

**Review by nashif — APPROVED** [review]

**Pull request merged**